DigitalLearningSolutions.Data/Services/RegistrationService.cs

```diff
@@ -150,21 +154,40 @@ public string RegisterDelegateByCentre(DelegateRegistrationModel delegateRegistr
                 );
             }
 
-            var supervisorDelegateRecordIds = supervisorDelegateService.GetPendingSupervisorDelegateRecordsByEmail(
+            var supervisorDelegateRecordIdsMatchingDelegate =
+                GetPendingSupervisorDelegateIdsMatchingDelegate(delegateRegistrationModel);
+
+            UpdateSupervisorDelegateRecordsWithNewCandidateId(
+                candidateNumber,
                 delegateRegistrationModel.Centre,
-                delegateRegistrationModel.Email
-            ).Select(record => record.ID).ToList();
+                supervisorDelegateRecordIdsMatchingDelegate
+            );
+
+            return candidateNumber;
+        }
 
-            if (supervisorDelegateRecordIds.Any())
+        private void UpdateSupervisorDelegateRecordsWithNewCandidateId(
+            string candidateNumber,
+            int centreId,
+            IEnumerable<int> recordIds
+        )
+        {
+            if (recordIds.Any())
             {
-                supervisorDelegateService.AddCandidateIdToSupervisorDelegateRecords(
-                    supervisorDelegateRecordIds,
-                    delegateRegistrationModel.Centre,
-                    candidateNumber
-                );
+                var delegateUser = userDataService.GetDelegateUserByCandidateNumber(candidateNumber, centreId)!;

```

Comment:
This ! is worrying me a bit. I don't think we should assume the candidate will be in the db because
- This method might be used by a future dev elsewhere, not straight after registration
- If there's a error connecting to the database, we'll get a NullReferenceException.
It would be nicer to handle the null case, log that we couldn't find the candidate in the database, and don't bother with adding them to the supervisor delegate records.
You could handle the error in here and change the name of the method from `Do this` to `Do this if candidate exists`. Or you could throw here and handle outside this method.